sql/core/src/main/scala/org/apache/spark/sql/RelationalGroupedDataset.scala

```diff
@@ -46,11 +47,13 @@ import org.apache.spark.sql.types.{NumericType, StructType}
  * @since 2.0.0
  */
 @Stable
-class RelationalGroupedDataset protected[sql](
-    private[sql] val df: DataFrame,
+class RelationalGroupedDataset[T] protected[sql](
```

Comment:
This is a stable API, is it OK to add type parameter? @srowen @dongjoon-hyun

Reply:
Although the goal seems to extend `Dataset[Row]` to `Dataset[T]` in 3.0.0, I'm not sure about this approach. Not only this change (the generic class), line 51 also changes from `df: DataFrame` to `ds: Dataset[T]`. If there is some 3rd party classes extending this, this variable name and type change will break those classes.

@viirya . Do we need this change for your original goal, `Add API ...`?

Reply:
BTW, it seems that MiMa doesn't complain about this change?

Reply:
another option is to add new API as `.as[(K, T)]`. It's not ideal as `T` should be known when we create the `RelationalGroupedDataset`, but we can avoid changing the stable API.

Reply:
I don't think it's a binary-incompatible change because of type erasure, so MiMa doesn't flag it. However it will indeed probably not be source-compatible. We should only do it if there's a pretty necessary reason in 3.0

Reply:
@dongjoon-hyun  > Do we need this change for your original goal, Add API ...?
To change from df: DataFrame to ds: Dataset[T] is because we need the typed information. For DataFrame, we have no idea about original object type.

Reply:
`.as[(K, T)]` sounds good so we don't need changing the stable API.

Reply:
This would break source compatibility.

Reply:
Just add `as[K, T]` and do not add typed parameter to `RelationalGroupedDataset` class, does it still break source compatibility?